### PR TITLE
fixed memory leak of ssl session reuse in dyups and session sticky module

### PR DESCRIPTION
**Pull request by chobits**
try to fix https://github.com/alibaba/tengine/issues/1016

* ~~todo: added test case for session reuse into dyups.t~~
* ~~todo: added test case for session reuse into session_sticky.t~~

### CODE REVIEW

**Comment by chobits** [comment]
try  patch: https://patch-diff.githubusercontent.com/raw/alibaba/tengine/pull/1708.patch firstly

**Comment by chobits** [comment]
```
ngx_http_dyups_save_peer_session
   -> ctx->ssl_session = ssl_session;
// ctx is alloced from r->pool, which get reference of ssl_sessoin. 
// After r is released, ssl_session is not freed .
//
// There is another problem, since the ssl_session exists on the current ctx of r.
// After new request r' comes in, there is no ssl session saved on the newly generated ctx of the new request r'.
// In such case,  ssl session reuse does not take effect
```

1. Session leaks 
2. Session reuse logic does not take effect.

File: modules/ngx_http_upstream_session_sticky_module/ngx_http_upstream_session_sticky_module.c [hunk 1]
```diff
@@ -246,16 +234,8 @@ ngx_http_upstream_session_sticky_init_peer(ngx_http_request_t *r,
     sspd->sscf = sscf;
     sspd->get_rr_peer = ngx_http_upstream_get_round_robin_peer;
 
-    r->upstream->peer.data = sspd;
```
Comment (chobits) [inline modules/ngx_http_upstream_session_sticky_module/ngx_http_upstream_session_sticky_module.c:249]: We assert that (sspd == &sspd->rr).
(refer to nginx core module: ngx_http_upstream_ip_hash_module.c)

File: modules/ngx_http_upstream_session_sticky_module/ngx_http_upstream_session_sticky_module.c [hunk 2]
```diff
@@ -246,16 +234,8 @@ ngx_http_upstream_session_sticky_init_peer(ngx_http_request_t *r,
...
     r->upstream->peer.get = ngx_http_upstream_session_sticky_get_peer;
 
-#if (NGX_HTTP_SSL)
-    r->upstream->peer.set_session =
-                            ngx_http_upstream_session_sticky_set_peer_session;
-    r->upstream->peer.save_session =
-                            ngx_http_upstream_session_sticky_save_peer_session;
-#endif
```
Comment (chobits) [inline modules/ngx_http_upstream_session_sticky_module/ngx_http_upstream_session_sticky_module.c:257]: Do not need to re-write set_session() and save_session() function when u->peer.data is point of ngx_http_upstream_rr_peer_data_t, see ngx_http_upstream_ip_hash_module.c.

Only when u->peer.data is changed to module private data, it needs to re-write these two callbacks. For example, see ngx_http_upstream_keepalive_module.c.

**Review by nandsky — COMMENTED** [review]
这个module里有不少code style问题，本次pr要不一起调整了？

**Review by nandsky — APPROVED** [review]
LGTM

**Review by wangfakang — APPROVED** [review]

**Comment by chobits** [comment]
> 这个module里有不少code style问题，本次pr要不一起调整了？

has been merged. May next time

**Pull request merged**